Mirko

mirkoRainer opened the pull request:
Really liked how @smartypan227 did the folder/namespace structure in his last PR so tried to capture that to the best of my ability. 
Keeping relevant items together hierarchically in the same namespace and folder. It aligns well with my vision for Domain Driven Design, hopefully it will translate to code well.

I would like to see this continue as the project moves forward. 
Thought from the other Collaborators?

mirkoRainer committed:
Created a Combat folder under Rules. Added relavent existing classes.
After seeing smartypan's folder restructure, I see a way forward in organizing. The intent is to capture relevant rules in the same folder tree/branch.

mirkoRainer committed:
Consolidating PC related classes and interfaces to align with folder structure.

mirkoRainer committed:
Putting all Creature related files within the Rules.Creature namespace and folder space.

mirkoRainer committed:
Created an Equipment namespace and relevant folder space.

mirkoRainer committed:
updated Using statement for new namespace/folder restructuring.

mirkoRainer committed:
Classes moved elsewhere with folder restructure.

mirkoRainer committed:
csproj update to reflect some empty folders that are part of the new folder/namespace structure.

mirkoRainer commented:
test comment for Slack Bot

smartypan227 commented on pf2-core/Core/Rules/Creature/Alignment.cs:
```diff
@@ -1,4 +1,4 @@
-namespace PF2.Core.Rules
+namespace PF2.Core.Rules.Creature
 {
     public enum Alignment
```
Just a thought about this class.  For success/failure we have a structure for isSuccess and isFailure.  This seems like a similar situation where we could have isGood or isEvil etc. for the purposes of condition checking for spell resolution.  Not sure which of the boolean / enum structures are better but maybe we want them to at least be consistent?

mirkoRainer replied:
That sounds like a good convention to use. Basically setting up flags for alignment, where if all of them are false, it would put you as true neutral.

smartypan227 replied:
My only concern is when it comes to spells and actions that have special abilities against creatures of a specific type.  Using an enum is simpler in most cases, except those where we'd have to do alignment = LawfulGood or NeutralGood or ChaoticGood for example.  I think the enum is probably cleaner but that may be me.

mirkoRainer replied:
It might not be easier when it comes to creating immunities and resistances. A spell targeting only Lawful creatures could have code like `if (immunity.Alignment.isLawful()) return Effect();`. that seems to read better than `if (immunity.Alignment == Alignment.Lawful)` (unless there's a more elegant way to write it).

smartypan227 replied:
In thinking about it, I think that the enum is cleaner when defining the 9 alignments and success/failure, and we could write extension functions like IsLawful(eAlignment alignment) or IsSuccess() to handle immunities and resistances.  when there's a set amount of possible combinations, an enum feels cleaner to me than multiple booleans.

rlugge replied:
Sorry to be late to the conversation, but I think this conversation raises the question of if we're engaging in YAML -- 'Ya ain't gonna need it'.  What is the intended scope of this project?  Character sheets, or a virtual tabletop?  Character Sheets are a much simpler project, a VT much, much more difficult.  We should take that discussion to slack, I think.

Either way, I firmly believe that enum is the right option here.  If you want .isEvil? or .isLawful methods, just define them as an extension on the enum.

https://stackoverflow.com/questions/5985661/methods-inside-enum-in-c-sharp

smartypan227 commented on pf2-core/pf2-core.csproj:
```diff
@@ -5,4 +5,8 @@
 
+  <ItemGroup>
+    <Folder Include="Core\Rules\Creature\Monster\" />
+    <Folder Include="Core\Rules\Creature\PlayerCharacter\NonPlayerCharacter\" />
```
NonPlayerCharacter seems like it should be its own folder inside the Creature Folder, not necessarily a subfolder of PlayerCharacter.  Also, what differentiates NonPlayerCharacter from Monster?

mirkoRainer replied:
I haven't read up on NPC and monster creation rules (might not even be completely out yet) but I know in past editions, there were NPC Classes that a monster might not have. Now a humanoid "monster" might have some NPC or even PC Classes. Again, this is making assumptions based on 1e and not anything I know of from 2e, which is why I stopped there. Doing "NonPcCreature" seems like a clunky name but I'm sure we can come up with something better with some thinking time.

rlugge replied:
Having a unified 'creature' folder makes sense to me.  Splitting NPCs out vs monsters does not.  What, exactly, goes into an NPC block that doesn't fit into a monster block?  And logically speaking, why would that go under the PC folder?

mirkoRainer replied:
Do we think that Creature.Monster.NonPlayerCharacter would make sense for folder structure? I do think that NPC's will have things that "monsters" won't but can't know until the GM guide comes out. I'll put NPC in Monster in Creature for now and we can make a better informed decision later. But I do agree that NPC shouldn't be in PC. :)

smartypan227 commented on pf2-core/Core/Rules/Combat/MeleeAttack.cs:
```diff
@@ -1,7 +1,8 @@
 using System.Collections.Generic;
 
-namespace PF2.Core.Rules
+namespace PF2.Core.Rules.Combat
```
Not super important but I believe the CRB calls it "Encounter Mode" not "Combat" but Combat is intuitive enough.

mirkoRainer replied:
True! Maybe Rules.EncounterMode.Combat and also Rules.EncounterMode.Social which will allow for separation of Social interaction logic and combat logic.
This train of thought also would mean having a Rules.ExplorationMode and Rules.DowntiemMode, which I like, since they have different logic.

smartypan227 replied:
That sounds perfect to me

mirkoRainer committed:
Arranging folder structure to make more sense. Currently the GM Guide is not out so the Creature/Monster/NPC folders will be hashed out more once it's released.

The three "modes" of play, Encounter/Exploration/Downtime now have a folderfor their respective log

Namespaces have been updated according to the folder structure.

mirkoRainer commented:
test.

mirkoRainer commented:
test
